src/compat/db_json.cpp

```diff
@@ -1168,6 +1361,339 @@ db_json_get_bool_as_str_from_value (const JSON_VALUE *doc)
+	    {
+	      result += sql_path_string[i++];
+	    }
+	  break;
+	// get object name from json, it can be .name or ."name"
+	case '.':
+	  if (sql_path_string[++i] == '\"')
+	    {
+	      i++;
+	      delimiter = '\"';
+	      // iterate until last quotes
+	      while (i < sql_path_length && sql_path_string[i] != '\"')
+		{
+		  // if found a special character append using its correspondent from map
+		  if (special_chars.find (sql_path_string[i]) != special_chars.end())
+		    {
+		      result += special_chars[sql_path_string[i]];
+		      if (result[0] != '#')
+			{
+			  result = "#" + result;
+			}
+		    }
+		  else
+		    {
+		      result += sql_path_string[i];
+		    }
+
+		  i++;
+		}
+	    }
+	  else
+	    {
+	      // if the name is not between quotes, it should start only with characters
+	      if ((sql_path_string[i] >= '0' && sql_path_string[i] <= '9')
+		  || sql_path_string[i] == '[' || i >= sql_path_length)
+		{
+		  return ER_JSON_INVALID_PATH;
+		}
+
+	      while (i < sql_path_length && sql_path_string[i] != '.' && sql_path_string[i] != '[')
+		{
+		  result += sql_path_string[i++];
+		}
+
+	      i--;
+	    }
+	  break;
+
+	default:
+	  return ER_JSON_INVALID_PATH;
+	}
+    }
+
+  json_pointer_out = result;
+  return NO_ERROR;
+}
+
+static void
+db_json_get_paths_helper (const JSON_VALUE &obj, const std::string &sql_path, std::vector<std::string> &paths)
+{
+  if (obj.IsArray())
+    {
+      int count = 0;
+
+      for (auto it = obj.GetArray().begin(); it != obj.GetArray().end(); ++it)
+	{
+	  std::stringstream ss;
+	  ss << sql_path << "[" << count++ << "]";
+	  db_json_get_paths_helper (*it, ss.str(), paths);
+	}
+    }
+  else if (obj.IsObject())
+    {
+      for (auto it = obj.MemberBegin(); it != obj.MemberEnd(); ++it)
+	{
+	  std::stringstream ss;
+	  ss << sql_path << '.' << '"' << it->name.GetString() << '"';
+	  db_json_get_paths_helper (it->value, ss.str(), paths);
+	}
+    }
+
+  paths.push_back (sql_path);
+}
+
+int
+db_json_get_all_paths_func (const JSON_DOC &doc, JSON_DOC *&result_json)
+{
+  JSON_POINTER p ("");
+  JSON_VALUE *head = const_cast<JSON_VALUE *> (reinterpret_cast<const JSON_VALUE *> (p.Get (doc)));
+  std::vector<std::string> paths;
+
+  db_json_get_paths_helper (*head, "$", paths);
+
+  result_json->SetArray();
+
+  for (auto &s : paths)
+    {
+      JSON_VALUE val;
+      val.SetString (s.c_str(), result_json->GetAllocator());
+      result_json->PushBack (val, result_json->GetAllocator());
+    }
+
+  return NO_ERROR;
+}
+
+int
+db_json_keys_func (const JSON_DOC &doc, JSON_DOC *&result_json, const char *raw_path)
+{
+  int error_code = NO_ERROR;
+  std::string json_pointer;
+
+  error_code = db_json_convert_sql_path_to_pointer (raw_path, doc, json_pointer);
+
+  if (error_code != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error_code, 0);
+      return error_code;
+    }
+
+  JSON_POINTER p (json_pointer.c_str());
+  JSON_VALUE *head = (JSON_VALUE *)p.Get (doc);
```

Comment:
I think this cast is superfluous

Reply:
in order to use the same style, i can use here the const_cast..

Reply:
I will make head variable const